feat(ingest/tableau): support ingestion of access roles

haeniya opened the pull request:
## Description
This PR enables the optional ingestion of access roles for the Tableau ingestion. To do this, we fetch the group permissions from the Tableau API, optionally transform it to the users needs and add them as roles to the Tableau asset. Currently, it is only implemented for Workbooks but could be extended to other assets as well.

The configuration in the recipe could look something like this:
```
source:
  type: tableau
  config:
    # Coordinates
    connect_uri: https://tableau.example.com

    access_role_ingestion:
      enable_workbooks: True
      role_prefix: "R-Tableau-"
      group_substring_start: 29
      role_description: "IAM Role required to access this Tableau asset."
      displayed_capabilities: ["Read", "Write", "Delete"]
      request_url: "https://iam.example.com/accessRequest?role=$ROLE_NAME"
      group_name_pattern:
        allow: [ "^.*_Consumer$" ]

    # Credentials
    username: "${TABLEAU_USER}"
    password: "${TABLEAU_PASSWORD}"

sink:
  type: "datahub-rest"
  config:
    server: "http://localhost:8080"
```

The `access` aspect was only available for datasets before. That's why I added it to containers now. The "Access Management" tab should only be visible in the UI if there's an access aspect associated with the container.

Let me know what you think and I will add some tests for this.

Thanks.

## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

githendrik commented on metadata-ingestion/src/datahub/ingestion/source/tableau.py:
```diff
@@ -297,6 +301,54 @@ def make_tableau_client(self, site: str) -> Server:
+    request_url: str = Field(
+        default="",
+        description="Request URL that is added to the role for requesting access. "
+        "You can use the variable $PERMISSION_NAME inside the URL string. "
```
nitpick: $PERMISSION_NAME or $ROLE_NAME ?

haeniya replied:
Good catch thanks. Change it.

haeniya commented on metadata-ingestion/src/datahub/ingestion/source/tableau.py:
```diff
@@ -264,8 +273,9 @@ def make_tableau_client(self, site: str) -> Server:
 
-            # From https://stackoverflow.com/a/50159273/5004662.
-            server._session.trust_env = False
+            if not self.session_trust_env:
```
This was something we discovered along the way. It seems to disable / bypass proxy settings which is not what we want in our case. That's why we made it configurable. Let me know what you think.

haeniya commented:
If anyone has an idea regarding the failing tests, please let me know. They mostly don't seem related to my changes. I see linting issues in unity source and failing tests for other sources. Am I missing something?

sid-acryl commented on metadata-ingestion/src/datahub/ingestion/source/tableau/tableau.py:
```diff
@@ -298,6 +308,54 @@ def make_tableau_client(self, site: str) -> Server:
+
+    displayed_capabilities: Optional[List[str]] = Field(
+        default=None,
+        description="Description of the role that will be displayed in the Access Management tab of Datahub.",
```
duplicate description for fields `role_description` and `displayed_capabilities`

sid-acryl commented on metadata-ingestion/src/datahub/ingestion/source/tableau/tableau.py:
```diff
@@ -298,6 +308,54 @@ def make_tableau_client(self, site: str) -> Server:
+        description="Use this property if you only need a substring of the group in Tableau to create the role used in the request_url.",
+    )
+
+    role_description: str = Field(
```
Do we need a map here as there might be multiple roles ?

haeniya replied:
For us it's currently enough to just have one generic description for all the roles. This could be extended to enable mapping between maybe a regex pattern for a role name and a role description in the future. But maybe we wait with this until we see the need for it, WDYT?

sid-acryl reviewed (changes requested):
Please update tableau connector documentation also for groups are getting ingested as role in DataHub with example of group name in Tableau and corresponding role in DataHub.

haeniya commented on metadata-ingestion/src/datahub/ingestion/source/tableau/tableau.py:
```diff
@@ -216,6 +220,11 @@ class TableauConnectionConfig(ConfigModel):
 
+    session_trust_env: bool = Field(
+        False,
+        description="Sets the trust_env property in the requests session to for example bypass proxy settings.",
```
WDYT about "Configures the trust_env property in the requests session. Setting it to false will bypass proxy settings."?

haeniya commented:
> Please update tableau connector documentation also for groups are getting ingested as role in DataHub with example of group name in Tableau and corresponding role in DataHub.

@sid-acryl do you mean adding a new section in https://github.com/datahub-project/datahub/blob/master/metadata-ingestion/docs/sources/tableau/tableau_pre.md that explains the new feature of ingesting access roles with some examples?

sid-acryl commented:
> > Please update tableau connector documentation also for groups are getting ingested as role in DataHub with example of group name in Tableau and corresponding role in DataHub.
> 
> @sid-acryl do you mean adding a new section in https://github.com/datahub-project/datahub/blob/master/metadata-ingestion/docs/sources/tableau/tableau_pre.md that explains the new feature of ingesting access roles with some examples?

Yup and please also update the `Concept Mapping`

haeniya commented:
I'm currently fixing the tests and adding a new one - just fyi.

sid-acryl commented on metadata-ingestion/docs/sources/tableau/tableau_pre.md:
````diff
@@ -69,6 +70,40 @@ Lineage is emitted as received from Tableau's metadata API for
+  config:
+    server: "http://localhost:8080"
+```
+Assuming you have groups in Tableau with names like `AB_XY00-Tableau-Access_A_123_PROJECT_XY_Consumer` or `AB_XY00-Tableau-Access_A_123_PROJECT_XY_Analyst` and corresponding IAM roles like `AR-Tableau-PROJECT_XY_Consumer` or `AR-Tableau-PROJECT_XY_Analyst`.
````
Please move this assumption to above the recipe and mention how to map this information in recipe to achieve correct transformation.

haeniya replied:
@sid-acryl I moved it above the recipe and improved the wording. Let me know what you think.

sid-acryl reviewed (changes requested):
Lets verify if we need any update in Quickstart guide of tableau (https://github.com/datahub-project/datahub/tree/master/docs/quick-ingestion-guides/tableau)

haeniya commented:
> Lets verify if we need any update in Quickstart guide of tableau (https://github.com/datahub-project/datahub/tree/master/docs/quick-ingestion-guides/tableau)

I don't think we need to update this part.

sid-acryl commented on metadata-ingestion/src/datahub/ingestion/source/tableau/tableau.py:
```diff
@@ -2796,6 +2864,18 @@ def emit_workbook_as_container(self, workbook: Dict) -> Iterable[MetadataWorkUni
+            self.config.access_role_ingestion
+            and self.config.access_role_ingestion.enable_workbooks
+        ):
+            logger.info(f"Ingest access roles of '{workbook.get(c.LUID)}'")
```
make it debug and log it as "workbook-id={....}"

haeniya replied:
Changed it. Do you want any other log messages to be debug?

sid-acryl reviewed (changes requested):
I am trying it locally, will add more review comments in coming days

sid-acryl commented on metadata-ingestion/src/datahub/ingestion/source/tableau/tableau.py:
```diff
@@ -3150,11 +3230,97 @@ def emit_site_container(self):
+                    description=self.config.access_role_ingestion.role_description,
+                )
+
+                logger.info(f"Upsert role '{role_urn}'.")
```
Lets make it debug

sid-acryl commented on metadata-ingestion/src/datahub/ingestion/source/tableau/tableau.py:
```diff
@@ -298,6 +308,56 @@ def make_tableau_client(self, site: str) -> Server:
+        description="Description of the role that will be displayed in the Access Management tab of Datahub.",
+    )
+
+    displayed_capabilities: Optional[List[str]] = Field(
```
I'm considering a suitable name for this configuration. How do you feel about `displayed_permissions` or `visible_permissions`?

haeniya replied:
In Tableau this is called a Capability:
![image](https://github.com/user-attachments/assets/43b4762f-387e-490f-813e-a5aec5ebff4d)
See: https://tableau.github.io/server-client-python/docs/api-ref or https://help.tableau.com/current/api/rest_api/en-us/REST/rest_api_ref_permissions.htm#delete_workbook_permission. I'd keep it as is but let me know what you think or if you want me to change it to something else.

sid-acryl commented on metadata-ingestion/src/datahub/ingestion/source/tableau/tableau.py:
```diff
@@ -449,6 +509,11 @@ class TableauConfig(
 
+    access_role_ingestion: Optional[AccessRolesIngestionConfig] = Field(
+        default=None,
+        description="Settings for configuration of access roles ingestion.",
```
Please update description as `Configuration settings for ingesting access roles. This allows Tableau groups to be imported as DataHub Roles.`

sid-acryl reviewed (changes requested):
What is the motivation behind below config properties 
 - role_prefix
 - group_substring_start
 - group_substring_end

haeniya commented:
> What is the motivation behind below config properties
> 
> * role_prefix
> * group_substring_start
> * group_substring_end

As I tried to describe it in the docs (https://github.com/datahub-project/datahub/pull/11157/files#diff-9a69800355baee2a7495dab637dd6bb46ba1f3be3a6b73bbc7bc70c22b9aa07fR78-R81) those are for transforming the group names in Tableau to what is used as role name in the request link. In our case we have IAM roles that are related to each Tableau group but the names are not exactly the same. These properties are currently enough for us to transform the group names coming from Tableau to their corresponding IAM roles that we want to use in the request link.

sid-acryl commented:
> > What is the motivation behind below config properties
> > 
> > * role_prefix
> > * group_substring_start
> > * group_substring_end
> 
> As I tried to describe it in the docs (https://github.com/datahub-project/datahub/pull/11157/files#diff-9a69800355baee2a7495dab637dd6bb46ba1f3be3a6b73bbc7bc70c22b9aa07fR78-R81) those are for transforming the group names in Tableau to what is used as role name in the request link. In our case we have IAM roles that are related to each Tableau group but the names are not exactly the same. These properties are currently enough for us to transform the group names coming from Tableau to their corresponding IAM roles that we want to use in the request link.

Are you referring to AWS IAM roles when you mention IAM?

haeniya commented:
> > > What is the motivation behind below config properties
> > > 
> > > * role_prefix
> > > * group_substring_start
> > > * group_substring_end
> > 
> > 
> > As I tried to describe it in the docs (https://github.com/datahub-project/datahub/pull/11157/files#diff-9a69800355baee2a7495dab637dd6bb46ba1f3be3a6b73bbc7bc70c22b9aa07fR78-R81) those are for transforming the group names in Tableau to what is used as role name in the request link. In our case we have IAM roles that are related to each Tableau group but the names are not exactly the same. These properties are currently enough for us to transform the group names coming from Tableau to their corresponding IAM roles that we want to use in the request link.
> 
> Are you referring to AWS IAM roles when you mention IAM?

Could be any Identity & Access Management tool. We're using Sailpoint but it doesn't really matter. It doesn't even need to be an IAM tool. Those fields are just there to transform the Tableau groups to what is used in the request link as `$ROLE_NAME`. Could be any tool. Maybe we remove IAM in the description to avoid confusion.

sid-acryl commented on metadata-ingestion/src/datahub/ingestion/source/tableau/tableau.py:
```diff
@@ -298,6 +308,56 @@ def make_tableau_client(self, site: str) -> Server:
+
+    request_url: str = Field(
+        default="",
+        description="Request URL that is added to the role for requesting access. "
```
As per code it looks like $ROLE_NAME is required to substitute the role name in request_url.

`Request URL that is added to access role metadata. The request URL must include $ROLE_NAME, For example 'https://example-host.com/order/$ROLE_NAME'`

haeniya replied:
No, `$ROLE_NAME` shouldn't be required. If it's not present it just won't be replaced. Or where do you see an issue?

sid-acryl reviewed (changes requested):
Let's remove the term IAM.
Is it possible for you to test it with other Identity or Cloud Provider (google, azure) ?

haeniya commented on metadata-ingestion/src/datahub/ingestion/source/tableau/tableau.py:
```diff
@@ -216,6 +220,11 @@ class TableauConnectionConfig(ConfigModel):
 
+    session_trust_env: bool = Field(
+        False,
+        description="Configures the trust_env property in the requests session. If set to false (default value) it will bypass proxy settings.",
```
Tbh, I'm not fully sure what exactly it does I just refer to the documentation here: https://requests.readthedocs.io/en/latest/api/#requests.Session.trust_env. Should I link this page in the description?

haeniya replied:
Added to the description of the field. Let me know if I should change anything.

haeniya commented on metadata-ingestion/src/datahub/ingestion/source/tableau/tableau.py:
```diff
@@ -265,8 +274,9 @@ def make_tableau_client(self, site: str) -> Server:
 
-            # From https://stackoverflow.com/a/50159273/5004662.
-            server._session.trust_env = False
+            if not self.session_trust_env:
```
Yep, we could do that I think. I just tried to not touch the old code if possible. But it makes more sense, I'll change it. We have a default of False so it should be fine.

haeniya commented on metadata-ingestion/src/datahub/ingestion/source/tableau/tableau.py:
```diff
@@ -298,6 +308,56 @@ def make_tableau_client(self, site: str) -> Server:
 
 
+class AccessRolesIngestionConfig(ConfigModel):
+    enable_workbooks: bool = Field(
```
The idea behind `enable_workbooks` is that we can add more entities in the future like views or dashboard. We could then just add a new config property `enable_dashboards` or `enable_views`. WDYT?

haeniya commented:
> Let's remove the term IAM. Is it possible for you to test it with other Identity or Cloud Provider (google, azure) ?

Changed various parts according to your feedback. Please have a look at the remaining comments. To your question - no, unfortunately, I don't have the option to test with other IAM providers. But as already said, it doesn't matter what IAM tool is used. The way access requests work in Datahub currently is that there is a link in the Access Management tab of an asset that redirects the user some external page. My changes are only related to this link that is created for each Tableau group. In theory, any link can be created if the role name can be derived from the Tableau group.

PS: I see a lot of tests failing due to recent changes with `TypeError: ContainerPropertiesClass.__init__() got an unexpected keyword argument 'env'`. Doesn't seem to be related to my changes. Could anyone have a look at that?

hsheth2 commented:
Hey folks, I should've jumped in a bit earlier on this PR.

I'm good with
- adding access aspect to containers
- all the relevant UI changes
- fetching user/group info for workbooks from the tableau API

The piece I am worried about is this - this convention feels a bit too specific to your setup.

```
Assuming you have groups in Tableau with names such as `AB_XY00-Tableau-Access_A_123_PROJECT_XY_Consumer` or `AB_XY00-Tableau-Access_A_123_PROJECT_XY_Analyst` and corresponding roles (e.g. in an IAM tool) like `AR-Tableau-PROJECT_XY_Consumer` or `AR-Tableau-PROJECT_XY_Analyst`.
Using the recipe below, would filter the groups that end with "_Consumer" and transform the group names into the corresponding roles.
With `group_start_index` and `group_end_index` you can define a substring of the group name to be used as role name and `role_prefix` can be used to add a prefix to the generated role name.
The role names then act as a substitute of `$ROLE_NAME` in the `request_url` and result in access request URLs like `https://iam.example.com/accessRequest?role=AR-Tableau-PROJECT_XY_Consumer`, for example.
```

If we can conclusively say that this is a standard convention for linking AWS / other IAM setups with Tableau, and that other DataHub users will also benefit from this specific setup, then I'm good with it. But I haven't really heard of this setup from other folks thus far.

At the same time, I do want to figure out how we can make this work for you. I'd personally be in favor of putting the user/group info in the container's custom properties. That way, you could still build a transformer that can convert custom properties into access aspects.

haeniya commented:
> Hey folks, I should've jumped in a bit earlier on this PR.
> 
> I'm good with
> 
> * adding access aspect to containers
> * all the relevant UI changes
> * fetching user/group info for workbooks from the tableau API
> 
> The piece I am worried about is this - this convention feels a bit too specific to your setup.
> 
> ```
> Assuming you have groups in Tableau with names such as `AB_XY00-Tableau-Access_A_123_PROJECT_XY_Consumer` or `AB_XY00-Tableau-Access_A_123_PROJECT_XY_Analyst` and corresponding roles (e.g. in an IAM tool) like `AR-Tableau-PROJECT_XY_Consumer` or `AR-Tableau-PROJECT_XY_Analyst`.
> Using the recipe below, would filter the groups that end with "_Consumer" and transform the group names into the corresponding roles.
> With `group_start_index` and `group_end_index` you can define a substring of the group name to be used as role name and `role_prefix` can be used to add a prefix to the generated role name.
> The role names then act as a substitute of `$ROLE_NAME` in the `request_url` and result in access request URLs like `https://iam.example.com/accessRequest?role=AR-Tableau-PROJECT_XY_Consumer`, for example.
> ```
> 
> If we can conclusively say that this is a standard convention for linking AWS / other IAM setups with Tableau, and that other DataHub users will also benefit from this specific setup, then I'm good with it. But I haven't really heard of this setup from other folks thus far.
> 
> At the same time, I do want to figure out how we can make this work for you. I'd personally be in favor of putting the user/group info in the container's custom properties. That way, you could still build a transformer that can convert custom properties into access aspects.

Hi @hsheth2, as discussed offline, I refactored the code and removed the access aspect and role creation again. The ingestion now only sets the Tableau groups as custom properties. Access and role creation will be done in a transformer. Please review the changes again and let me know what you think. Thanks.

hsheth2 commented on metadata-ingestion/docs/sources/tableau/tableau_pre.md:
```diff
@@ -57,6 +57,7 @@ This ingestion source maps the following Source System Concepts to DataHub Conce
 | User                        | [User (a.k.a CorpUser)](../../metamodel/entities/corpuser.md) | Optionally Extracted              |
 | Workbook                    | [Container](../../metamodel/entities/container.md)            | SubType `"Workbook"`              |
 | Tag                         | [Tag](../../metamodel/entities/tag.md)                        | Optionally Extracted              |
+| Permissions (Groups)        | [Role](../../metamodel/entities/role)                         | Optionally Extracted              |
```
let's update this

hsheth2 commented on metadata-ingestion/src/datahub/ingestion/source/tableau/tableau.py:
```diff
@@ -3168,11 +3214,63 @@ def emit_site_container(self):
+        for group in TSC.Pager(self.server.groups):
+            self.group_map[group.id] = group
+
+    def _filter_capabilities(self, capabilities):
```
let's get some type annotations on this. I think this should be returning an `optional[list[str]]`, and the formatting stuff should be done by the caller

hsheth2 commented on metadata-ingestion/src/datahub/ingestion/source/tableau/tableau.py:
```diff
@@ -3168,11 +3214,63 @@ def emit_site_container(self):
+
+        groups = []
+        for rule in permissions:
+            if rule.grantee.tag_name == "group":
```
are groups the only possibilities? or can individual users also get grants - and that's something that's just not supported by this PR?

haeniya replied:
In theory, there could be individual users that have access to the Tableau assets. However, from a request access perspective only groups really make sense to display in the central catalog, that's why it's currently not supported by this PR.

hsheth2 commented on metadata-ingestion/src/datahub/ingestion/source/tableau/tableau.py:
```diff
@@ -3168,11 +3214,63 @@ def emit_site_container(self):
+                )
+                groups.append(f"{group.name}{capabilities}")
+
+        return {"groups": ", ".join(groups)} if len(groups) > 0 else None
```
might be easier to just `json.dumps(groups)`

haeniya replied:
So would you rather want to see the groups with their capabilities as a JSON string in the UI? Would be easier to parse later in the transformer but could be a bit hard on the eyes in the custom properties.

haeniya replied:
It now looks like this: ![image](https://github.com/user-attachments/assets/d36febc2-1a81-4368-8878-3c750f901ab4)

hsheth2 commented on metadata-ingestion/src/datahub/ingestion/source/tableau/tableau.py:
```diff
@@ -298,6 +304,29 @@ def make_tableau_client(self, site: str) -> Server:
+
+    displayed_capabilities: Optional[List[str]] = Field(
+        default=None,
+        description="List of Tableau permission capabilities (e.g. Read, Write, Delete) that will be displayed for each role in the custom property. "
```
I'm kinda curious about the motivation behind this config option? My initial instinct was that we should just ingest all capabilities, without any filtering

haeniya replied:
The initial motivation was that there's a lot of very specific capabilities in Tableau: https://help.tableau.com/current/server-linux/en-us/permissions_capabilities.htm. We thought this could be a bit confusing for users in the catalog that's why we added the possibility to only display certain capabilities in Datahub.

Since we're now moving most of the logic into a custom transformer now, I'm fine with removing this and showing all capabilities. Let me what you think.

hsheth2 replied:
yeah let's do that

hsheth2 commented on metadata-ingestion/src/datahub/ingestion/source/tableau/tableau.py:
```diff
@@ -298,6 +304,29 @@ def make_tableau_client(self, site: str) -> Server:
             ) from e
 
 
+class GroupIngestionConfig(ConfigModel):
```
`PermissionIngestionConfig`?

it feels a bit strange to be super specific to groups, when this is broadly about access controls and permissions

hsheth2 commented:
@haeniya feels like we're pretty close here :)

hsheth2 commented:
@haeniya looks like the tests are failing
```
FAILED tests/integration/tableau/test_tableau_ingest.py::test_group_ingestion - datahub.ingestion.run.pipeline.PipelineInitError: Failed to configure the source (tableau): 1 validation error for TableauConfig
group_ingestion
  extra fields not permitted (type=value_error.extra)
```
